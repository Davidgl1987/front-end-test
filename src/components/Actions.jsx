import { useRef } from 'react'
import { toast } from 'react-hot-toast'
import { CartIcon } from '../icons'
import { useStore } from '../services'
import { useCart } from '../hooks'
import { OptionSelector } from './OptionSelector'

export const Actions = ({ product }) => {

  const cartProduct = useRef({
    product,
    color: 0,
    storage: 0
  })

  const { addToCart } = useStore(state => state)

  const { mutateAsync } = useCart()

  const handleOnChange = ({ target }) => {
    cartProduct.current = { ...cartProduct.current, [target.name]: target.value }
  }

  const onAddToCart = () => {
    const { color, storage, product } = cartProduct.current
    const colorCode = product.options.colors[color].code
    const storageCode = product.options.storages[storage].code
    mutateAsync({ id: product.id, colorCode, storageCode })
      .then(response => {
        console.log({ response })
        addToCart(cartProduct.current)
        toast.success('Producto añadido correctamente')
      })
      .catch(error => {
        toast.error(`Error: ${error.message}`)
      })
  }

  return (
    <div className='rounded-xl bg-white shadow-lg mb-4'>
      <ul>
        <li className='w-full px-4 py-2 sm:grid sm:grid-cols-2 sm:gap-x-4'>
          <OptionSelector
            label='Color'
            name={'color'}
            onChange={handleOnChange}
            selected={cartProduct.colorCode}
            options={product.options.colors}
          />
          <OptionSelector
            label='Almacenamiento'
            name={'storage'}
            onChange={handleOnChange}
            selected={cartProduct.storageCode}
            options={product.options.storages}
          />
          <button
            className='w-full p-2 mb-2 text-lg font-bold text-white text-center inline-flex items-center justify-center bg-cyan-700 rounded-lg hover:bg-cyan-800 focus:ring-4 focus:outline-none focus:ring-cyan-300 transition-all sm:col-span-2'
            onClick={onAddToCart}
          >
            <CartIcon className='text-white w-4 h-4 mr-2' /> Añadir al carrito
          </button>
        </li>
      </ul>
    </div>
  )
}
