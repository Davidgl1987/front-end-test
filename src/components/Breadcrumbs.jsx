import { Link } from 'react-router-dom'
import { HomeIcon, SeparatorIcon } from '../icons'

export const Breadcrumbs = ({ breadcrumbs }) => {
  return (
    <nav className='flex py-4' aria-label='Breadcrumb'>
      <ol className='inline-flex items-center space-x-1 md:space-x-3'>
        {breadcrumbs.map((breadcrumb, i) => (
          <li key={i}>
            <Link to={breadcrumb.url} className='inline-flex items-center text-sm font-medium text-gray-600 hover:text-cyan-600 dark:text-gray-400 dark:hover:text-white'>
              {i === 0 ? <HomeIcon /> : <SeparatorIcon />}
              <div className='ml-1 text-sm font-medium text-gray-600 hover:text-cyan-600 md:ml-2 dark:text-gray-400 dark:hover:text-white'>{breadcrumb.name}</div>
            </Link>
          </li>
        ))}
      </ol>
    </nav>
  )
}
