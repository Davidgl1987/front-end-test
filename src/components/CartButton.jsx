import { CartIcon } from '../icons'
import { useStore } from '../services/store'

export const CartButton = () => {
  const cart = useStore(state => state.cart)
  return (
    <>
      <button
        className='bg-cyan-700 hover:bg-cyan-800 transition-all hover:scale-110 focus:ring-4 focus:outline-none focus:ring-cyan-300 font-medium rounded-full text-sm p-2.5 text-center inline-flex items-center mr-2 relative'
        type='button'
      >
        <CartIcon className='text-white' />
        {cart.length > 0 && <div className='absolute inline-flex items-center justify-center w-5 h-5 text-xs text-white bg-red-500 rounded-full -top-2 -right-2'>{cart.length}</div>}
      </button>
    </>
  )
}
