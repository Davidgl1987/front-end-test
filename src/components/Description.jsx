import { Fragment } from 'react'

export const Description = ({ product }) => {
  return (
    <div className='rounded-xl bg-white shadow-lg mb-4'>
      <ul>
        <li className='w-full px-4 py-2 border-b border-gray-200'>
          <h3 className='mb-2 text-2xl font-bold text-gray-900'>{product.model}</h3>
          <p className='mb-2 text-gray-500'>{product.brand}</p>
        </li>
        <li className='w-full px-4 py-2 border-b border-gray-200'>
          <h2 className='mb-2 text-3xl font-extrabold tracking-tight text-cyan-700'>{product.price}€</h2>
        </li>
        <li className='w-full border-gray-200 text-sm'>
          <table>
            <tbody>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>CPU</td>
                <td className='px-3'>{product.cpu}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>RAM</td>
                <td className='px-3'>{product.ram}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>Sistema operativo</td>
                <td className='px-3'>{product.os}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>Resolución de pantalla</td>
                <td className='px-3'>{product.displayResolution}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>Batería</td>
                <td className='px-3'>{product.battery}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>Cámara principal</td>
                <td className='px-3'>{product.primaryCamera.map((cam, i) => <Fragment key={i}>{cam}&nbsp;</Fragment>)}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>Cámara secundaria</td>
                <td className='px-3'>{product.secondaryCmera.map((cam, i) => <Fragment key={i}>{cam}&nbsp;</Fragment>)}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1'>Dimensiones</td>
                <td className='px-3'>{product.dimentions}</td>
              </tr>
              <tr>
                <td className='bg-gray-100 px-3 py-1 rounded-bl-xl'>Peso</td>
                <td className='px-3'>{product.weight}</td>
              </tr>
            </tbody>
          </table>
        </li>
      </ul>
    </div>
  )
}
