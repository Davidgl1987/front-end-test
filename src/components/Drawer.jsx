import { useStore } from '../services/store'

export const Drawer = () => {
  const { cart, previewCart, togglePreviewCart } = useStore(state => state)

  return (
    <aside
      className={
        'fixed overflow-hidden z-10 bg-gray-900 bg-opacity-25 inset-0 transform ease-in-out ' +
        (previewCart
          ? 'transition-opacity opacity-100 duration-300 translate-x-0'
          : 'transition-all opacity-0 delay-300 translate-x-full')
      }
    >
      <section
        className={
          'w-screen max-w-lg right-0 absolute bg-cyan-50 h-full shadow-xl delay-200 duration-300 ease-in-out transition-all transform ' +
          (previewCart ? 'translate-x-0' : 'translate-x-full')
        }
      >
        <article className='relative w-screen max-w-lg pt-16 pb-10 flex flex-col overflow-y-scroll h-full'>
          <header className='p-4 font-bold text-lg text-center'>Carrito</header>
          <div className='p-4 pt-0'>
            {cart.length === 0 && <div className='text-center'>No has a√±adido productos a la cesta</div>}
            {cart.length > 0 && cart.map(({ product, color, storage }) => (
              <ul className='mb-2' key={product.id}>
                <li className='flex justify-center px-4 py-6 hover:bg-cyan-200 rounded-xl'>
                  <img className='w-16 h-16' src={product.imgUrl} alt='Bonnie image' />
                  <div className='pl-6 flex flex-col justify-center'>
                    <div className='text-gray-800 text-md font-semibold mb-1.5 dark:text-gray-400'>
                      {product.brand} {product.model}
                    </div>
                    <div className='text-xs text-gray-500'>{product.options.colors[color].name} - {product.options.storages[storage].name}</div>
                  </div>
                </li>
              </ul>))}
            {/* <pre className='text-red-300'>{JSON.stringify(cart, null, 2)}</pre> */}
          </div>
        </article>
      </section>
      <section
        className='w-screen h-full cursor-pointer'
        onClick={() => togglePreviewCart()}
      />
    </aside>
  )
}
