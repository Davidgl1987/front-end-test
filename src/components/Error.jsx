export const Error = ({ text }) => {
  return (
    <div className='flex items-center justify-center p-4 mb-4 text-sm text-red-800 border-4 border-red-300 rounded-lg bg-red-50 dark:bg-gray-800 dark:text-red-400 dark:border-red-800' role='alert'>
      <div>
        <span className='font-medium'>Â¡Error!</span> {text}
      </div>
    </div>

  )
}
