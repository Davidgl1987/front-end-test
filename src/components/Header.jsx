import { Link } from 'react-router-dom'
import { CartButton } from './'

export const Header = () => {
  return (
    <header>
      <nav className='bg-cyan-600 text-white fixed w-full z-20 top-0 left-0 shadow-md'>
        <div className='max-w-screen-xl flex flex-wrap items-center justify-between mx-auto p-4'>
          <Link to='/' className='flex items-center hover:text-cyan-800 transition-all'>
            <img src='/mobile-store.svg' className='h-8 mr-3 hover:scale-110 transition-all' alt='Front-end text logo' />
            <span className='self-center text-2xl font-semibold whitespace-nowrap dark:text-white'>Front-end test</span>
          </Link>
          <CartButton />
        </div>
      </nav>
    </header>
  )
}
