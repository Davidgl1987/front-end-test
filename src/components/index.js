export * from './Header'
export * from './CartButton'
export * from './ProductItem'
export * from './Search'
export * from './Actions'
export * from './Description'
export * from './Image'
export * from './Breadcrumbs'
export * from './Loader'
export * from './Error'
export * from './Drawer'
