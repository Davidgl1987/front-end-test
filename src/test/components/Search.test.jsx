import { render, screen, fireEvent } from '@testing-library/react'
import { test, describe, expect, vi } from 'vitest'
import { Search } from '../../components'

describe('Search', () => {
  const onChange = vi.fn()
  const text = 'Samsung'

  test('Snapshot', () => {
    const { container } = render(<Search onChange={onChange} />)
    expect(container).toMatchSnapshot()
  })
  test('Fire onChange event on input change', () => {
    render(<Search onChange={onChange} />)
    const input = screen.getByRole('textbox')
    fireEvent.input(input, { target: { value: text } })
    expect(input.value).toBe(text)
    expect(onChange).toBeCalledTimes(1)
    expect(onChange).toBeCalledWith(text)
  })
})
