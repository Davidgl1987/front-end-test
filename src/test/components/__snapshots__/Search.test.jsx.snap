// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Search > Snapshot 1`] = `
<div>
  <div
    class="relative"
  >
    <div
      class="absolute inset-y-0 left-0 flex items-center pl-3 pointer-events-none"
    >
      <svg
        aria-hidden="true"
        class="text-gray-600"
        fill="none"
        height="1em"
        viewBox="0 0 20 20"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="m19 19-4-4m0-7A7 7 0 1 1 1 8a7 7 0 0 1 14 0Z"
          stroke="currentColor"
        />
      </svg>
    </div>
    <input
      class="w-full p-4 pl-10 text-sm text-gray-900 rounded-xl shadow-lg bg-gray-50 focus:ring focus:ring-cyan-600 focus:border-cyan-500 focus-visible:outline-none transition-all"
      placeholder="Acer, Tab, Alcatel..."
      type="text"
    />
  </div>
</div>
`;
